Encoders!

ky28059 opened the pull request:
### Resources
- [Talon sensor guide](https://docs.ctre-phoenix.com/en/latest/ch14_MCSensor.html)
- [`WPI_TalonSRX` docs (most methods to do with encoders are in `TalonSRX` or `BaseMotorController` which `WPI_TalonSRX` inherits)](https://www.ctr-electronics.com/downloads/api/java/html/classcom_1_1ctre_1_1phoenix_1_1motorcontrol_1_1can_1_1_w_p_i___talon_s_r_x.html)
- [Example projects on GitHub](https://github.com/CrossTheRoadElec/Phoenix-Examples-Languages/tree/master/Java%20General)
- [This thread on Chief Delphi](https://www.chiefdelphi.com/t/using-encoder-with-talon-srx/145483/18)

ky28059 committed:
Initial testing

proan commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -17,16 +17,20 @@
         public static final int fLeftMotorPort = 15;
-        public static final int fRightMotorPort = 1;
         public static final int bLeftMotorPort = 14;
+
```
extra line?

ky28059 committed:
Add wheel circumference to hopefully calculate distance

ky28059 committed:
Add position getting/setting, fix constants

ky28059 committed:
Slow drive speed, zero sensors

ky28059 committed:
Start odometry threading

ky28059 committed:
Odometry package and class

This commit will be overridden by using WPI's builtin odometry classes, but I am pushing it to keep it in history.

ky28059 committed:
Use WPI odometry class

ky28059 committed:
Small changes

ky28059 committed:
navX AHRS

ky28059 committed:
Merge left and right sensor methods to use odometry class

ky28059 commented:
### Resources for the odometry portion of this PR
- [navX website](https://pdocs.kauailabs.com/navx-mxp/examples/) and [`AHRS` docs](https://www.kauailabs.com/public_files/navx-mxp/apidocs/java/com/kauailabs/navx/frc/AHRS.html)
- `DifferentialDrivePoseEstimator` [docs](https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/estimator/DifferentialDrivePoseEstimator.html) and [guide](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/state-space/state-space-pose_state-estimators.html)

ky28059 committed:
Change `new Pose2d()` to `getRobotPosition()`

ky28059 committed:
Measure the constant!

This can be tested and tuned over time

ky28059 committed:
Bind A button to zero position

e3l commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -42,6 +43,7 @@
 
   // Controllers
   private XboxController controlXbox = new XboxController(0);
+  private JoystickButton xboxAButton = new JoystickButton(controlXbox, XboxController.Button.kA.value);
```
why not use the controller directly, is that not possible

ky28059 replied:
https://docs.wpilib.org/en/stable/docs/software/commandbased/binding-commands-to-triggers.html
https://first.wpi.edu/FRC/roborio/beta/docs/java/edu/wpi/first/wpilibj/XboxController.html#getAButtonPressed--
I think the `JoystickButton` class just wraps the `XboxController`'s button methods so you can do event based "when pressed" stuff instead of having to check periodically whether the button was pressed using `getAButtonPressed()` and then going from there. Je crois que setting up button bindings via `JoystickButton` is the reason for the `controllerBindings` method.

e3l commented on src/main/java/frc/robot/commands/tank/DriveTankCommand.java:
```diff
@@ -7,41 +7,40 @@
 
   @Override
   public boolean isFinished() {
     return true;
```
big lemon

e3l commented on src/main/java/frc/robot/commands/tank/DriveTankCommand.java:
```diff
@@ -7,41 +7,40 @@
-    tankSubsystem.setCarDrivePowers(yScale, angularScale, squareInput);
+    // TODO: the zeroing of the sensors takes a non-zero amount of time and `isFinished()` returns 
+    // true before the sensors are zeroed.
+    // Every other run of this command will fail, with the second run serving only to zero the sensors 
```
can't we save the starting position of the encoders and simply add to it?

also, for serious movement, we should just take the position from the poseestimator instead of doing encoder logic ourselves. we can probably put .distanceToPoint(), etc. in a wrapper for all localization stuff

proan replied:
You should consider a localization system for the future.

e3l commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -20,6 +30,14 @@
 
+  private final AHRS ahrs;
+
+  private final DifferentialDrivePoseEstimator odometry;
```
in another comment I mentioned "all localization stuff"... we should take all the localization logic and put it in its own class and run it on its own thread. logically and from my experience, dead reckoning is more accurate the less time each loop is -- no matter your odometry method (e.g. circular), you're making an assumption about what happens between each tick, so reducing the tick interval reduces the fudginess of your guess. 

if we use the periodic() method, the commandbased system actively waits until an interval is hit before calling periodics, it's something like 20ms by default? i can't remember

proan replied:
I second the localization system approach.

e3l commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -131,4 +182,24 @@ public void setTankDrivePowers(double leftScale, double rightScale) {
+    rightMain.setSelectedSensorPosition(0);
+
+    // https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/estimator/DifferentialDrivePoseEstimator.html#resetPosition(edu.wpi.first.wpilibj.geometry.Pose2d,edu.wpi.first.wpilibj.geometry.Rotation2d)
+    odometry.resetPosition(new Pose2d(), Rotation2d.fromDegrees(ahrs.getAngle()));
```
why not reset the gyro

ky28059 replied:
https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/estimator/DifferentialDrivePoseEstimator.html#resetPosition(edu.wpi.first.wpilibj.geometry.Pose2d,edu.wpi.first.wpilibj.geometry.Rotation2d)
> The gyroscope angle does not need to be reset here on the user's robot code. The library automatically takes care of offsetting the gyro angle.

And the angle was reset to 0 during testing so it seems to work.

ky28059 committed:
Change constant to inches, remeasure

proan commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -70,7 +72,7 @@ public RobotContainer() {
 
     // Instantiate commands
-    tankCommand = new DriveTankCommand(tankSubsystem, 0, 0);
+    tankCommand = new DriveTankCommand(tankSubsystem, 12);
```
Is this "12" related to the "12 inches" comment on L117?
I don't love magic numbers, but a comment here would be helpful if you don't want to make this a well-named constant.

ky28059 replied:
Yeah, the idea was for it to drive forward "12 inches" in autonomous to then measure that distance and check if the constant was working properly, but the different speeds on each side prevented that from really happening. I'll add a comment to this.

proan commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -20,6 +30,12 @@
+
+  private final DifferentialDrivePoseEstimator odometry;
+
+  public static final double ENCODER_TICKS_TO_INCHES = 32 / 142.40;
```
Fortunately you have the `.40` in this equation, otherwise the result is `0`. Ask me how I know and how many hours I spent hunting for this bug.

If you have integer division, you get an integer result.

proan commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -44,18 +60,51 @@ public TankSubsystem() {
+    double leftDistance = leftMain.getSelectedSensorPosition() * ENCODER_TICKS_TO_INCHES;
+    double rightDistance = rightMain.getSelectedSensorPosition() * ENCODER_TICKS_TO_INCHES;
+
+    odometry.update(gyroAngle, wheelVelocities, leftDistance, rightDistance);
```
Seems like this interface isn't great. We're passing in a `DifferentialDriveWheelSpeeds` object and then separate positions for left and right wheels. These two interfaces should match. I know this isn't your interface, but keep this in mind.

proan commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -44,18 +60,51 @@ public TankSubsystem() {
+
+    odometry.update(gyroAngle, wheelVelocities, leftDistance, rightDistance);
+
+    System.out.println(odometry.getEstimatedPosition());
```
Maybe print out a little text for context in the logs/terminal?

ky28059 committed:
Add better documentation

ky28059 committed:
Remove joysticks from encoder branch

ky28059 committed:
Readd odometry thread

Wasn't able to test this that much because the drivetrain became possessed by ghosts, but can test next week on the actual drivetrain.

ky28059 commented:
[`CANSparkMax`](https://www.revrobotics.com/content/sw/max/sw-docs/java/com/revrobotics/CANSparkMax.html) has a similar API to `WPI_TalonSRX`, but some notable differences:

----

![image](https://user-images.githubusercontent.com/60120929/146609919-e9a8d800-2761-4511-a3f7-109e03a884f8.png)
Followers automatically inherit their leader's inversion (and can invert their leader's inversion if needed).

![image](https://user-images.githubusercontent.com/60120929/146610047-c63798a8-5d48-4c19-9611-360460b60dab.png)
Percent output motor setting is the default.

![image](https://user-images.githubusercontent.com/60120929/146610712-51fc10b6-06a5-45f8-ad2a-5f0381f912a6.png)
The type of motor (brushed or brushless) must be given in the constructor.

![image](https://user-images.githubusercontent.com/60120929/146611726-155ee6d3-2476-417b-a3ce-964388b2d076.png)
Brushless hall encoders can be accessed via `getEncoder()` (but this is only for temporary testing, as on the build robot we'll use external magencoders)

ky28059 committed:
Talon -> Spark Max

ky28059 committed:
Implement preliminary path following command

See comment in FollowPathCommand

ky28059 commented on src/main/java/frc/robot/subsystems/tank/TankSubsystem.java:
```diff
@@ -125,4 +125,16 @@ public void setTankDrivePowers(double leftScale, double rightScale) {
+  public void setMotorVoltages(double left, double right) {
+    // TODO: stub on purpose, see comment in FollowPathCommand
+    // This doesn't really match our current interface, but neither does the other option;
+    // Should a state on this subsystem be added to switch between voltage and percent output?
```
Alternatively, instead of `tankSubsystem` containing states and continually setting the motor controller output in `periodic`, the motor controllers can be the ones to store the state and methods can just set their values. Instead of
```java
private int leftPower;
private int rightPower;

public void setTankDrivePowers(double leftScale, double rightScale) {
  this.leftPower = leftScale;
  this.rightPower = rightScale;
}

@Override
public void periodic() {
  // These must be in percent output mode
  leftMain.set(leftPower);
  rightMain.set(rightPower);
}
```
why not have
```java
public void setTankDrivePowers(double leftScale, double rightScale) {
  leftMain.set(leftScale);
  rightMain.set(rightScale);
}
```
? This would also allow for more flexibility with the output mode, such as
```java
public void setTankDrivePowers(double leftScale, double rightScale) {
  leftMain.set(leftScale);
  rightMain.set(rightScale);
}

public void setTankDriveVoltages(double leftVoltage, double rightVoltage) {
  leftMain.setVoltage(leftVoltage);
  rightMain.setVoltage(rightVoltage);
}
```

ky28059 replied:
Allowing voltages would also resolve 1/2 of the current problems with the path following command.

ky28059 committed:
Remove states from `tankSubsystem`

ky28059 commented:
It seems like WPILib's path/trajectory things *must* be in meters (see [PathWeaver export unit](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/pathweaver/creating-pathweaver-project.html) and all the WPILib trajectory docs) so I will convert the odometry to that as well. It seems like PathWeaver can have its interface be in feet or inches while exporting paths as meters, which should make the forced meterization of everything less painful.

ky28059 committed:
Inches -> meters

ky28059 committed:
Add secondary constructor to FollowPathCommand

PinkBlueSky committed:
Fix turning direction; remove CAN ID 0

PinkBlueSky committed:
Edit to functional drivetrain state

ky28059 committed:
Measure some constants

Prepared to do https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-characterization/characterization-routine.html#running-the-characterization-routine and measure the rest on Sunday, then perhaps test finally!

ky28059 committed:
Measure more constants!

These will probably have to be remeasured with the Robot Characterization tool. When testing pathfinding with these, the robot immediately swerves sharply to the left (which it shouldn't be doing). It might have something to do with the right encoder readings being inverted during characterization (left reports 4 meters, right reports -4), but I've tried setting various options in the config to try and fix that to no avail. Will definitely look into it further at some point.

ky28059 committed:
Revert constants, invert gyro, invert left motors

ky28059 committed:
Fix inverted gyro error in resetPosition

ky28059 committed:
New constants!

Recharacterized using process described in https://www.chiefdelphi.com/t/the-future-of-pathweaver/393014/82. `Kp` might need to be tuned, as does `ENCODER_TICKS_TO_METERS` most likely.

ky28059 committed:
Update WPILib and vendor packages to 2022

Also push experimental changes to PID constants and auton testing, which still doesn't quite work

ky28059 committed:
Add preliminary shuffleboard code

ky28059 committed:
Move odometry to `tankSubsystem`

With the new REV version, CANSparkMaxes with the same can ID cannot be created (it throws errors which kill the robot program)

ky28059 committed:
Measure new constants with sysid

Hopefully these ones work better!

ky28059 committed:
`Field2d`, new constants

ky28059 committed:
Cleanup, remove elevator subsystem

ky28059 commented:
### Closing(ish) thoughts
A lot of things made sense when I realized what `RamseteCommand` and associated WPILib path following things actually did -- given a trajectory, the path following generates a list of left and right **velocities** for the drivetrain to follow the trajectory. My confusion from the PID oscillation stemmed from the assumption that the PID loop in the command was for *position*, so an oscillation would surely cause the robot to move back and forth; it never did because the PID loop was for *velocity* instead (this is why `kI` and `kD` were set to 0 in the example; accelerating to a target velocity is a use case for just a P controller). Oscillations due to a high `kP` were between 1.9m/s and 2.1m/s and not 1.9m and 2.1m so there was no back and forth driving. Furthermore, the confusion on why the robot stopped before the end position when constants were misconfigured can also be explained by realizing that the list of velocities probably ended, with the constants being incorrect leading to the robot not following the velocities properly and ending in an incorrect spot.

A lot of these confusions could probably have been resolved sooner had I read the docs for [`RamseteController`](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/trajectories/ramsete.html#getting-adjusted-velocities) a little more carefully.

ky28059 committed:
Change angle entry to use a gyro complex widget

eggaskin reviewed (approved):
Some constants to be fine tuned/changed, but Kevin's computer is not working so first merge.

ky28059 commented:
> Some constants to be fine tuned/changed, but Kevin's computer is not working so first merge.

I don't think constants need to be changed, I just wanted to test shuffleboard (which wasn't cooperating). I think launching it from driver station as opposed to vscode fixed the issue, so the branch can probably be deleted.
